python/lsst/daf/butler/core/dimensions/graph.py

```diff
@@ -0,0 +1,473 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("DimensionGraph", "DimensionConfig")
+
+import itertools
+from ..config import ConfigSubset
+from .sets import DimensionSet, toNameSet
+from .elements import Dimension, DimensionJoin
+
+
+class DimensionConfig(ConfigSubset):
+    component = "dimensions"
+    requiredKeys = ("version", "units", "joins")
+    defaultConfigFile = "dimensions.yaml"
+
+
+class DimensionGraph:
+    r"""A automatically-expanded collection of both `Dimension`\s and
+    `DimensionJoin`\s.
+
+    A `DimensionGraph` behaves much like a `DimensionSet` containing only
```

Comment:
Move these paragraphs to a [Notes](
https://developer.lsst.io/python/numpydoc.html#notes) section.